File: node/fuzz/targets/endpoint-json.mjs
/**
 * `VerifyReceiptEndpoint.verifyReceiptJson`, the one entry point that takes a
 * request body rather than a receipt: JSON parse, `receipt-data` extraction,
 * the receipt-base64 rule, then the whole DER path.
 *
 * Its documented contract is stronger than the other targets' — it never
 * throws at all — so that is what is asserted: any body, any bytes, gets a
 * JSON object with a numeric `status` back. The typed result behind that
 * body is checked too: exactly one of receipt and failureReason, the same
 * status, and never INTERNAL_ERROR. That reason means an unexpected error
 * inside the pipeline, or content a trusted signer signed that the library
 * cannot read; a fuzzer cannot forge a trusted signature, so for fuzz input
 * it can only be the first, and that is a bug.
 */
import { VerifyReceiptEndpoint } from '../../dist/index.js';
import { RECEIPT_ANCHORS, asUtf8 } from '../harness.mjs';

const endpoint = new VerifyReceiptEndpoint({
  trustedRoots: RECEIPT_ANCHORS,
  environment: 'Sandbox',
});

export function fuzz(data) {
  const body = asUtf8(data);
  if (body === null) {
    return;
  }
  let response;
  try {
    response = endpoint.verifyReceiptJson(body);
  } catch (error) {
    throw new Error(
      `the endpoint threw ${error?.constructor?.name}: ${error?.message}, but it documents that it never throws`,
      { cause: error },
    );
  }
  let parsed;
  try {
    parsed = JSON.parse(response);
  } catch (error) {
    throw new Error(`the endpoint answered with something that is not JSON: ${response}`, {
      cause: error,
    });
  }
  if (parsed === null || typeof parsed !== 'object' || typeof parsed.status !== 'number') {
    throw new Error(`the endpoint answered without a numeric status: ${response}`);
  }

  let result;
  try {
    result = endpoint.verifyReceiptResult(body);
  } catch (error) {
    throw new Error(
      `verifyReceiptResult threw ${error?.constructor?.name}: ${error?.message}, but it documents that it never throws`,
      { cause: error },
    );
  }
  if ((result.receipt === null) === (result.failureReason === null)) {
    throw new Error('verifyReceiptResult broke its receipt/failureReason invariant');
  }
  if (result.failureReason === 'INTERNAL_ERROR') {
    throw new Error('verifyReceiptResult hit an internal error', { cause: result.failureCause });
  }
  if (result.status !== parsed.status) {
    throw new Error(`verifyReceiptResult status ${result.status} differs from ${response}`);
  }
}

File: python/fuzz/targets/endpoint_json.py
"""``VerifyReceiptEndpoint.verify_receipt_json``, the one entry point that
takes a request body rather than a receipt: JSON parse, ``receipt-data``
extraction, the receipt-base64 rule, then the whole DER path.

Its documented contract is stronger than the other targets' — it never raises
at all — so that is what is asserted: any body, any bytes, gets a JSON object
with a numeric ``status`` back.
"""

import json

from harness import ENDPOINT, InvariantViolation, as_text, run


def one_input(data: bytes) -> None:
    body = as_text(data)
    if body is None:
        return
    try:
        response = ENDPOINT.verify_receipt_json(body)
    except Exception as error:
        raise InvariantViolation(
            f"the endpoint raised {type(error).__name__}: {error}, "
            "but it documents that it never raises"
        ) from error
    try:
        parsed = json.loads(response)
    except ValueError as error:
        raise InvariantViolation(
            f"the endpoint answered with something that is not JSON: {response!r}"
        ) from error
    status = parsed.get("status") if isinstance(parsed, dict) else None
    if not isinstance(status, int) or isinstance(status, bool):
        raise InvariantViolation(f"the endpoint answered without a numeric status: {response!r}")


if __name__ == "__main__":
    run(one_input)
